MaxText/standalone_checkpointer.py

```diff
@@ -0,0 +1,107 @@
+"""
+ Copyright 2023 Google LLC
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
+# pylint: disable=g-bad-todo, abstract-method, consider-using-with, ungrouped-imports
+"""Standalone checkpointer - only saves and restores checkpoints at regular intervals, accesses storage needs."""
+
+# Calling jax.device_count here prevents a "TPU platform already registered" error.
+# See github.com/google/maxtext/issues/20 for more
+
+import datetime
+import os
+
+from typing import Sequence
+from absl import app
+from flax.linen import partitioning as nn_partitioning
+import jax
+import numpy as np
+
+import checkpointing
+import max_utils
+import max_logging
+import pyconfig
+from train import setup_mesh_and_model, get_first_step, validate_train_config
+
+from layers import models
+
+from jax.experimental.compilation_cache import compilation_cache as cc
+
+Transformer = models.Transformer
+
+def checkpoint_loop(config, state=None):
+  """Main Checkpointing loop.
+  Saves checkpoints.
+  Args:
+    config:
+    state:
+    ckpt_path:
+  Returns:
+  """
+  init_rng, writer, checkpoint_manager, mesh, model, _, tx = setup_mesh_and_model(config)
+
+  unboxed_abstract_state, state_mesh_annotations = max_utils.get_abstract_state(model, tx,
+                                                config, init_rng, mesh, is_training=True)
+  # A barrier to sync all hosts before starting to restore checkpoint
+  jax.experimental.multihost_utils.sync_global_devices("dummy")
+  checkpoint_load_start = datetime.datetime.now()
+  with nn_partitioning.axis_rules(config.logical_axis_rules):
+    state, _ = checkpointing.load_state_if_possible(checkpoint_manager,
+                                                config.load_parameters_path,
+                                                config.load_full_state_path,
+                                                unboxed_abstract_state,
+                                                mesh,
+                                                state_mesh_annotations)
+  jax.experimental.multihost_utils.sync_global_devices("dummy")
```

Comment:
No a sync doesn't work here -- you need a block_until_ready. Sync makes sure everyone is executing the same python, block_until_ready means the tensor is actually ready. Jax will by default not worry about it.

Also, you need to use unique strings for all your different sync_global_devices -- not super important in this case, but you don't want to accidentally sync different barriers.

Reply:
Thanks for explaining, done!